src/battle/overlay_12_0224E4FC.c

```diff
@@ -1962,3 +1965,920 @@ void ov12_02251710(BattleSystem *bsys, BATTLECONTEXT *ctx) {
+    } else if (moveTypeDefault) {
+        moveType = moveTypeDefault;
+    } else {
+        moveType = ctx->unk_334.moveData[moveNo].type;
+    }
+    
+    movePower = ctx->unk_334.moveData[moveNo].power;
+
+    //STAB
+    if (!(ctx->linkStatus & (1 << 11)) && (GetBattlerVar(ctx, battlerIdAttacker, BMON_DATA_TYPE_1, NULL) == moveType || GetBattlerVar(ctx, battlerIdAttacker, BMON_DATA_TYPE_2, NULL) == moveType)) {
+        if (GetBattlerAbility(ctx, battlerIdAttacker) == ABILITY_ADAPTABILITY) {
+            damage *= 2;
+        } else {
+            damage = damage * 15 / 10;
+        }
+    }
+    
+    if (CheckBattlerAbilityIfNotIgnored(ctx, battlerIdAttacker, battlerIdTarget, ABILITY_LEVITATE) == TRUE && moveType == TYPE_GROUND && itemTarget != HOLD_EFFECT_SPEED_DOWN_GROUNDED) {
+        *moveStatusFlags |= MOVE_STATUS_LEVITATE_IMMUNE;
+    } else if (ctx->battleMons[battlerIdTarget].unk88.magnetRiseTurns && !(ctx->battleMons[battlerIdTarget].moveEffectFlags & MOVE_EFFECT_INGRAIN) && moveType == TYPE_GROUND && itemTarget != HOLD_EFFECT_SPEED_DOWN_GROUNDED) {
+        *moveStatusFlags |= MOVE_STATUS_MAGNET_RISE_IMMUNE;
+    } else {
+        i = 0;
+        do {
+            if (sTypeEffectiveness[i][0] == TYPE_FORESIGHT) {
+                if ((ctx->battleMons[battlerIdTarget].status2 & STATUS2_FORESIGHT) || GetBattlerAbility(ctx, battlerIdAttacker) == ABILITY_SCRAPPY) {
+                    break;
+                } else {
+                    i++;
+                    continue;
+                }
+            }
+            if (sTypeEffectiveness[i][0] == moveType) {
+                u8 monType = sTypeEffectiveness[i][1];
+                if (monType == GetBattlerVar(ctx, battlerIdTarget, BMON_DATA_TYPE_1, NULL) && ov12_02251C74(ctx, battlerIdAttacker, battlerIdTarget, i) == TRUE) {
+                    damage = ov12_022583B4(ctx, battlerIdAttacker, sTypeEffectiveness[i][2], damage, movePower, moveStatusFlags);
+                }
+                if (monType == GetBattlerVar(ctx, battlerIdTarget, BMON_DATA_TYPE_2, NULL) && GetBattlerVar(ctx, battlerIdTarget, BMON_DATA_TYPE_1, NULL) != GetBattlerVar(ctx, battlerIdTarget, BMON_DATA_TYPE_2, NULL) && ov12_02251C74(ctx, battlerIdAttacker, battlerIdTarget, i) == TRUE) {
+                    damage = ov12_022583B4(ctx, battlerIdAttacker, sTypeEffectiveness[i][2], damage, movePower, moveStatusFlags);
+                }
+            }
+            i++;
+        } while (sTypeEffectiveness[i][0] != TYPE_ENDTABLE);
+    }
+    
+    if (CheckBattlerAbilityIfNotIgnored(ctx, battlerIdAttacker, battlerIdTarget, ABILITY_WONDER_GUARD) == TRUE && ov12_02258440(ctx, moveNo) &&
+        (!(*moveStatusFlags & MOVE_STATUS_SUPER_EFFECTIVE) || ((*moveStatusFlags & MOVE_STATUS_ANY_EFFECTIVE) == MOVE_STATUS_ANY_EFFECTIVE)) && movePower) {
+        *moveStatusFlags |= MOVE_STATUS_WONDER_GUARD_IMMUNE;
+    } else if (!(ctx->linkStatus & (1 << 11)) && !(ctx->linkStatus & (1 << 15))) {
```

Comment:
constants? though I think you mentioned that you were going to do the link status one later... not 100% sure though